### PR TITLE
Service controller

### PR DESCRIPTION
**Pull request by lena-aiu**
Hi John, please review the changes in service_controller.rb, where I tried to create an access controller. I'm not sure that it's 100% right, because for example used (admin? || intern?) instead of (current_user.role.include?("administrator") || current_user.role.include?("intern")) @jrmcgarvey

### CODE REVIEW

**Review by jrmcgarvey — COMMENTED** [review]
This is basically correct, except for the one error identified below, but I think there is significant risk in merging this PR so soon before the demo, so I will not merge it right now.  Please fix the error so we can merge later.  I will also add that the code is kind of verbose -- one could reduce some of the logic and if/else statements.

File: app/controllers/services_controller.rb [hunk 1]
```diff
@@ -53,13 +100,35 @@ def create
       redirect_to @service
     else
       flash.now.alert = @service.errors.full_messages.to_sentence
-      render :edit
+      render :new
```
[the remark this answers is not in the document]
Comment (lena-aiu) [reply app/controllers/services_controller.rb]: Okay, thank you

**Comment by lena-aiu** [comment]
I just push changes that I made base on your comments @jrmcgarvey

**Comment by lena-aiu** [comment]
@jrmcgarvey Please review the code. I fixed errors with the request tests for services, making changes in the access controller for services. Also fixed an error with the request test for categories, uncommenting the index method in categories_controller.rb.

File: app/controllers/categories_controller.rb [hunk 2]
```diff
@@ -61,18 +61,18 @@ def create
...
     end
-    @category = Category.new(category_params) 
+    @category = Category.new(category_params)
     if @category.save
       flash.notice = "The customer record was created successfully."
       redirect_to @category
     else
       flash.now.alert = @category.errors.full_messages.to_sentence
-      render :new  
+      render :new
```
Comment (jrmcgarvey) [inline app/controllers/categories_controller.rb:75]: This shoudl be render :edit

File: spec/requests/services_spec.rb [hunk 3]
```diff
@@ -105,7 +105,7 @@
       service_attributes.delete(:name)
       expect { post services_path, params: {service: service_attributes}
     }.to_not change(Service, :count)
-      expect(response).to render_template(:edit)
+      expect(response).to render_template(:new)
```
Comment (jrmcgarvey) [inline spec/requests/services_spec.rb:108]: this should be render_template(:edit)

File: app/controllers/services_controller.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (jrmcgarvey) [reply app/controllers/services_controller.rb:80]: This is still render :new it looks like.  It should be render :edit

**Review by jrmcgarvey — CHANGES_REQUESTED** [review]
There are three places where we have :new where it should be :edit.

File: app/controllers/services_controller.rb [hunk 4]
```diff
@@ -13,57 +13,88 @@ class ServicesController < ApplicationController
...
+        marker.lng service.longitude
+        marker.infowindow "<a href='https://www.google.com/maps/dir/Current+Location/#{service.address}' target='_blank'>#{service.name}</a>"
+      end
   end
 
   # GET /serivices/new
   def new
-    @service = Service.new
+    # @service = Service.new
+    if  nil? || volunteer?
```
Comment (jrmcgarvey) [inline app/controllers/services_controller.rb:37]: The if nil? here is not correct.  I think what is meant is, if current_user.role.nil? .  If nil? just means, is the current instance of the services_controller nil, which is always false.  In any case, these three lines are unnecessary.  You should take them out and just start with if admin? || intern?

File: app/controllers/services_controller.rb [hunk 5]
```diff
@@ -13,57 +13,88 @@ class ServicesController < ApplicationController
...
+      flash.notice = "You are not authorized for that operation."
+      redirect_to home_index_path
+    end
+
   end
 
   # GET /services/1/edit
   def edit
-
+    if  nil? || volunteer?
```
Comment (jrmcgarvey) [inline app/controllers/services_controller.rb:51]: Again, the if nil? is not correct, and these 3 lines ae not needed.

File: app/controllers/services_controller.rb [hunk 6]
```diff
@@ -13,57 +13,88 @@ class ServicesController < ApplicationController
...
+    else
+      flash.notice = "You are not authorized for that operation."
+      redirect_to home_index_path
+    end
   end
 
   # POST /services
   # POST /services.json
   def create
+    if  nil? || volunteer?
```
Comment (jrmcgarvey) [inline app/controllers/services_controller.rb:65]: again, the if ni? is not correct, and these three lines are unneeded.

File: app/controllers/services_controller.rb [hunk 7]
```diff
@@ -13,57 +13,88 @@ class ServicesController < ApplicationController
...
       flash.now.alert = @service.errors.full_messages.to_sentence
       render :edit
     end
+
   end
 
   # PATCH/PUT /customers/1
   # PATCH/PUT /customers/1.json
   def update
+    if  nil? || volunteer?
```
Comment (jrmcgarvey) [inline app/controllers/services_controller.rb:88]: again, the if nil? is not correct, and these three lines are unneeded.

File: app/controllers/services_controller.rb [hunk 8]
```diff
@@ -75,6 +106,15 @@ def update
   # DELETE /customers/1
   # DELETE /customers/1.json
   def destroy
+    if  nil? || volunteer?
```
Comment (jrmcgarvey) [inline app/controllers/services_controller.rb:109]: same deal with the if nil?

**Review by jrmcgarvey — CHANGES_REQUESTED** [review]
Elena, I apologize, but I did find some more changes that are really needed.  These are marked in the review.  The main problem is the use of if nil?  which is not correct.

File: app/controllers/services_controller.rb (hunk 4 above)
Comment (lena-aiu) [reply app/controllers/services_controller.rb]: Okay, I will make changes

File: app/controllers/services_controller.rb (hunk 5 above)
Comment (lena-aiu) [reply app/controllers/services_controller.rb]: Okay, I will make changes

File: app/controllers/services_controller.rb (hunk 7 above)
Comment (lena-aiu) [reply app/controllers/services_controller.rb]: Okay, I will make changes

**Comment by lena-aiu** [comment]
@jrmcgarvey Hi John, I removed an unnecessary part of the code with if nil?, also changed the look of condition for admin and intern. Please review it. Thank you

**Review by jrmcgarvey — APPROVED** [review]
looks ok!

**Pull request merged**